### PR TITLE
Add public share to VDAF API

### PR DESCRIPTION
**Pull request by divergentdave**
This updates the VDAF API to include a public share along with per-aggregator input shares. A new associated type, `PublicShare`, is added to the `Vdaf` trait, and the signatures of `Client::shard()` and `Aggregator::prepare_init()` are updated to plumb the public share through. The public share will be required to implement Poplar1 as specified in drafts 02 and 03, but it is not used yet in the codebase. (The Poplar paper's IDPF requires a public share, but ToyIdpf does not)

This depends on #286.

### CODE REVIEW

**Comment by divergentdave** [comment]
I'm going to add another `#[allow(clippy::type_complexity)]` to `shard()`, as it's only triggering a warning on 1.58, not 1.63.

File: src/vdaf/poplar1.rs
Comment (cjpatton) [inline src/vdaf/poplar1.rs:345]: For consistency
```suggestion
    type PublicShare = (); // TODO: Replace this when the IDPF from [BBCGGI21] is implemented.
```

**Review by cjpatton — APPROVED** [review]
LGTM, though given that we are once again complicating the API with something that Prio3 doesn't use, I wonder if it's worth revisiting our strategy here. An alternative to modifying `Vdaf` et al would be to create a new set of traits that consume/produce the public_share. 

The updside of this would be reducing deadcode. I can see a couple of downsides:
1. We would no longer be able to write code that is generic for both Prio3 and Poplar1.
2. We would have to write different testing frameworks for different VDAF variants, leading to lots of redundant code and probably increasing maintenance cost.

Before merging, I think it's worth spending a few minutes pondering if these downsides can be mitigated. Calling on @tgeoghegan and @branlwyd for thoughts they might have here.

**Review by tgeoghegan — APPROVED** [review]

**Comment by tgeoghegan** [comment]
I think it's important to maintain a strong correspondence between this crate and the VDAF specification, because if we find that something is ugly, difficult to implement in `prio` or difficult for clients of `prio` to use, then that suggests a problem with the VDAF spec (or let's say an opportunity to improve the VDAF spec).

My hope is that in the future, new VDAFs will be defined and implemented in crates outside of `libprio-rs`, and then DAP implementations that are generic over `trait Vdaf` should be able to use those new VDAFs with no code changes (except those needed to map a configuration value to a particular `Vdaf` instantiation).

If DAP implementations wind up having to write different code to contend with `trait Vdaf` and `trait VdafWithPublicParam`, then I contend that the VDAF specification is not achieving its goal of abstracting over VDAFs, and ought to separately specify a VDAF and a VDAFWithPublicParam (much like how it contains distinct definitions of DAFs and VDAFs).

File: src/vdaf/poplar1.rs
Comment (glandium) [inline src/vdaf/poplar1.rs:361]: The API would have been clearer if `()` had been `PublicShare`.

Comment (tgeoghegan) [reply src/vdaf/poplar1.rs:361]: Good point! https://github.com/divviup/libprio-rs/pull/371

**Pull request merged**
